presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkNativeTaskRdd.java

```diff
@@ -0,0 +1,185 @@
+
+import static com.facebook.presto.spark.classloader_interface.ScalaUtils.emptyScalaIterator;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static scala.collection.JavaConversions.asJavaCollection;
+import static scala.collection.JavaConversions.seqAsJavaList;
+
+/**
+ * PrestoSparkTaskRdd represents execution of Presto stage, it contains:
+ * - A list of shuffleInputRdds, each of the corresponding to a child stage.
+ * - An optional taskSourceRdd, which represents ALL table scan inputs in this stage.
+ * <p>
+ * Table scan is present when joining a bucketed table with an unbucketed table, for example:
+ * Join
+ * /  \
+ * Scan  Remote Source
+ * <p>
+ * In this case, bucket to Spark partition mapping has to be consistent with the Spark shuffle partition.
+ * <p>
+ * When the stage partitioning is SINGLE_DISTRIBUTION and the shuffleInputRdds is empty,
+ * the taskSourceRdd is expected to be present and contain exactly one empty partition.
+ * <p>
+ * The broadcast inputs are encapsulated in taskProcessor.
+ */
+public class PrestoSparkNativeTaskRdd<T extends PrestoSparkTaskOutput>
+        extends PrestoSparkTaskRdd<T>
+{
+    public static <T extends PrestoSparkTaskOutput> PrestoSparkNativeTaskRdd<T> create(
+            SparkContext context,
+            Optional<PrestoSparkTaskSourceRdd> taskSourceRdd,
+            // fragmentId -> RDD
+            Map<String, RDD<Tuple2<MutablePartitionId, PrestoSparkMutableRow>>> shuffleInputRddMap,
+            PrestoSparkTaskProcessor<T> taskProcessor)
+    {
+        requireNonNull(context, "context is null");
+        requireNonNull(taskSourceRdd, "taskSourceRdd is null");
+        requireNonNull(shuffleInputRddMap, "shuffleInputRddMap is null");
+        requireNonNull(taskProcessor, "taskProcessor is null");
+        ImmutableList.Builder<String> shuffleInputFragmentIds = ImmutableList.builder();
+        ImmutableList.Builder<RDD<Tuple2<MutablePartitionId, PrestoSparkMutableRow>>> shuffleInputRdds = ImmutableList.builder();
+        for (Map.Entry<String, RDD<Tuple2<MutablePartitionId, PrestoSparkMutableRow>>> entry : shuffleInputRddMap.entrySet()) {
+            shuffleInputFragmentIds.add(entry.getKey());
+            shuffleInputRdds.add(entry.getValue());
+        }
+        return new PrestoSparkNativeTaskRdd<>(context, taskSourceRdd, shuffleInputFragmentIds.build(), shuffleInputRdds.build(), taskProcessor);
+    }
+
+    @Override
+    public Iterator<Tuple2<MutablePartitionId, T>> compute(Partition split, TaskContext context)
+    {
+        PrestoSparkTaskSourceRdd taskSourceRdd = getTaskSourceRdd();
+        List<Partition> partitions = seqAsJavaList(((ZippedPartitionsPartition) split).partitions());
+        int expectedPartitionsSize = (taskSourceRdd != null ? 1 : 0) + getShuffleInputRdds().size();
+        if (partitions.size() != expectedPartitionsSize) {
```

Comment:
instead of if (!condition) throw exception, use Gauva 

checkState (condition, "error message", parameters)